### ISSUE TITLE
`convox start` host port conflicts

### ISSUE THREAD

**Issue by mattmanning** [convox/rack#484]
One of our users has a `docker-compose.yml` that looks something like:

```
foo:
  ports:
    - 80:3000
bar:
  ports:
    - 80:3000
```

This works fine in production because foo and bar will each have their own ELB, but when running locally with `convox start` there is a conflict on host port 80.

How can we change our tooling to improve the local development experience in this case?

**Comment by nzoschke** [comment]
I believe in the newest version of Docker Engine and Docker Compose you can use networks to get around this. You can make two networks for this app, and assign each process to an isolated network and access them with a different name/ip but same port.

**Comment by nzoschke** [comment]
Here are some possible paths I see here.

We strongly support and encourage this in production. We spend time to make the development environment support the pattern.

We strongly don't support it at all. We add validation to `convox start` and `convox deploy` to not build the app in this case.

We take a middle ground. It's possible in production but not simple in development. You can use 2 docker-compose files, one for dev and one for deploy if you want this.

--

How does this work with docker-compose? Can we copy it's implementation? Docker networking should be flexible enough to make this work locally.

**Comment by nzoschke** [comment]
Copying @mattmanning's additional information in #573:

I have a `docker-compose.yml` like this:

```yaml
web:
  links:
    - db1
    - db2
db1:
  ports:
    - 3306
db2:
  ports:
    - 3306
```

The DB image I'm using wants to listen on port 3306 (MySQL). This configuration should be fine in production because each will get its own load balancer. However, locally with `convox start` both containers are started with `3306:3306` as the port mapping, causing a conflict on host port 3306.

Since I'm using linking I don't really care what the host port is. I think this problem could be solved by choosing a random port for it in both cases.

**Comment by mwarkentin** [comment]
Once you start needing to run multiple apps, you can run into the same issues locally:

`app1/docker-compose.yml`:

```yaml
web:
  - ports: 443:80
```

`app2/docker-compose.yml`L

```yaml
web:
  - ports: 443:80
```

You'll get a host port conflict on `443` if you try running both apps at once w/ `convox start`.

**Comment by mwarkentin** [comment]
Here's a bit more information on our current set up that we're migrating to Convox, and what we're trying to achieve for local dev and in production.

### Local dev (current):

* We have ~15 apps with web processes configured in Vagrant. Each of these is configured to run on a unique port from `8000-8015`.
* We have DNS records set up for `wave-local.com` (`my.wave-local.com`, `accounting.wave-local.com`, etc.)
  * All of these domains point to `33.33.33.100`, which is where the vagrant private network is configured (`config.vm.network :private_network, ip: "33.33.33.100"`)
* nginx is configured inside of Vagrant to send the request to the correct port based on the domain: eg. `my.wave-local.com` -> `port 9000`

```
                                      ┌────────────────────────────────────────────────┐
                                      │                                                │
                                      │                      DNS                       │
                                      │        *.wave-local.com -> 33.33.33.100        │
                                      │                                                │
                                      └────────────────────────────────────────────────┘
                                                               │
                                                               │
                                                               │
                                                               │
          Vagrant                                              │
                                                               │
┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─│─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
                                                               ▼                                                               │
│                                                  ┌───────────────────────┐
                                                   │                       │                                                   │
│                                                  │                       │
                                                   │         nginx         │                                                   │
│                                                  │                       │
                                                   │                       │                                                   │
│                                                  └───────────────────────┘
                     ┌─────────────────┐            ┌─────┐    │   ┌─────┐       ┌─────────────────────────┐                   │
│             ┌──────┤my.wave-local.com├──────┬─────┤ ... ├────┴───┤ ... ├───┬───┤ receipts.wave-local.com ├──┐
              │      └─────────────────┘      │     └─────┘        └─────┘   │   └─────────────────────────┘  │                │
│             ▼                               ▼                              ▼                                ▼
   ┌────────────────────┐          ┌────────────────────┐         ┌────────────────────┐           ┌────────────────────┐      │
│  │                    │          │                    │         │                    │           │                    │
   │                    │          │                    │         │                    │           │                    │      │
│  │                    │          │                    │         │                    │           │                    │
   │                    │          │                    │         │                    │           │                    │      │
│  │        8000        │          │        8001        │         │        ...         │           │        8015        │
   │     (identity)     │          │    (accounting)    │         │                    │           │     (receipts)     │      │
│  │                    │          │                    │         │                    │           │                    │
   │                    │          │                    │         │                    │           │                    │      │
│  │                    │          │                    │         │                    │           │                    │
   │                    │          │                    │         │                    │           │                    │      │
│  └────────────────────┘          └────────────────────┘         └────────────────────┘           └────────────────────┘
                                                                                                                               │
└ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
```

### Staging / production (current)

* Not really important, just a standard VM based set up with DNS based routing to our different apps

### Local dev (convox)

* With one app we're fine, we just configured the port forwarding `443:80`
* When we added a second app with the same configuration, we had a port conflict on `443` on our Host
* I started looking into switching to a range-based port mapping (like we currently do in Vagrant), but that effected our convox rack deployments negatively (see below)
* For now, we're setting `443:80` for all of our apps, but this only works because they're isolated - other apps which communicate with each other will need to run at the same time, so can't have a host port conflict

### Staging / production (convox racks)

* All requests should go through HTTPS (port 443)
* Perhaps there's a way to use a different port? CloudFlare wouldn't let us point our domain at a specific port, only at the ELB domain - and we don't want to expose port numbers in the URL for our apps
* Works great with `443:80` - all apps are available at `https://<app ELB domain>`

### Diagrams

Here's what happens if we use `443:80` for multiple apps:

```
                                                 ┌────────────────────────────────────────────────────────────┐        ┌────────────────────────────────────────────────────────────┐
                                                 │           **                                   **          │        │                   **     **       **  ********             │
                                                 │          /**                                  /**          │        │                  ****   /**      /** **//////              │
                                                 │          /**        ******   *****   ******   /**          │        │                 **//**  /**   *  /**/**                    │
                                                 │          /**       **////** **///** //////**  /**          │        │                **  //** /**  *** /**/*********             │
                                                 │          /**      /**   /**/**  //   *******  /**          │        │               **********/** **/**/**////////**             │
                                                 │          /**      /**   /**/**   ** **////**  /**          │        │              /**//////**/**** //****       /**             │
                                                 │          /********//****** //***** //******** ***          │        │              /**     /**/**/   ///** ********              │
                                                 │          ////////  //////   /////   //////// ///           │        │              //      // //       // ////////               │
                                                 └────────────────────────────────────────────────────────────┘        └────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────┐           ┌────────────────────────┐        ┌────────────────────────┐                                         ┌─ ── ── ── ── ── ── ── ── ──
│app1/docker-compose.yml              │           │         Proxy          │        │          web           │                                         │          Instance          │
│                                     │           │                        │        │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │         ┌────────────────────────┐      │ ┌────────────────────────┐ │
│web:                                 │         ┌─┴─┐                    ┌─┴─┐    ┌─┴─┐   ┌──┐               │         │          ELB           │        │          web           │
│  ports:                             │         │443├────────────────────▶rnd├────▶rnd├───▶80│   nginx     │ │ ┌───────┴──────────────┐         │      │ │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │ │
│    - 443:80                         │         └─┬─┘                    └─┬─┘    └─┬─┘   └──┘               │ │ https://app1.elb.com │       ┌─┴─┐    ├─┴─┐   ┌──┐               │ │
│  labels:                            │           │                        │        │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │ │   (no port req'd)    │───────▶rnd├────▶rnd│───▶80│   nginx     │ │
│    - convox.port.443.protocol=https │           │                        │        │                        │ └───────┬──────────────┘       └─┬─┘    ├─┬─┘   └──┘               │ │
└─────────────────────────────────────┘           └────────────────────────┘        └────────────────────────┘         │                        │      │ │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │ │
                                                                                                                       │                        │        │                        │
                                                                                                                       └────────────────────────┘      │ └────────────────────────┘ │
                                                                                                                                                       └ ── ── ── ── ── ── ── ── ── ┘

                                                                                                                                                       ┌─ ── ── ── ── ── ── ── ── ──
                                                                                                                                                       │          Instance          │
┌─────────────────────────────────────┐           ┌────────────────────────┐        ┌────────────────────────┐         ┌────────────────────────┐      │ ┌────────────────────────┐ │
│app2/docker-compose.yml              │           │         Proxy          │        │          web           │         │          ELB           │        │          web           │
│                                     │       ┌───┴────┐                   │        │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │ ┌───────┴──────────────┐         │      │ │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │ │
│web:                                 │       │  port  │                 ┌─┴─┐    ┌─┴─┐   ┌──┐               │ │ https://app2.elb.com │       ┌─┴─┐    ├─┴─┐   ┌──┐               │ │
│  ports:                             │       │conflict├─────────────────▶rnd├────▶rnd├───▶80│   nginx     │ │ │   (no port req'd)    ├───────▶rnd├────▶rnd│───▶80│   nginx     │ │
│    - 443:80                         │       │  443   │                 └─┬─┘    └─┬─┘   └──┘               │ └───────┬──────────────┘       └─┬─┘    ├─┬─┘   └──┘               │ │
│  labels:                            │       └───┬────┘                   │        │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │         │                        │      │ │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │ │
│    - convox.port.443.protocol=https │           │                        │        │                        │         │                        │        │                        │
└─────────────────────────────────────┘           └────────────────────────┘        └────────────────────────┘         └────────────────────────┘      │ └────────────────────────┘ │
                                                                                                                                                       └ ── ── ── ── ── ── ── ── ── ┘


                                                            ┌─────────────────────────────────────┐                                ┌─────────────────────────────────────┐
                                                            │                                     │                                │                                     │
                                                            │               #     #               │                                │               #     #               │
                                                            │               ##    #               │                                │                #   #                │
                                                            │               # #   #               │                                │                 # #                 │
                                                            │               #  #  #               │                                │                  #                  │
                                                            │               #   # #               │                                │                  #                  │
                                                            │               #    ##               │                                │                  #                  │
                                                            │               #     #               │                                │                  #                  │
                                                            │                                     │                                │                                     │
                                                            └─────────────────────────────────────┘                                └─────────────────────────────────────┘

```

Here's what happens if we use a separate port for each app:

```
                                                 ┌────────────────────────────────────────────────────────────┐        ┌────────────────────────────────────────────────────────────┐
                                                 │           **                                   **          │        │                   **     **       **  ********             │
                                                 │          /**                                  /**          │        │                  ****   /**      /** **//////              │
                                                 │          /**        ******   *****   ******   /**          │        │                 **//**  /**   *  /**/**                    │
                                                 │          /**       **////** **///** //////**  /**          │        │                **  //** /**  *** /**/*********             │
                                                 │          /**      /**   /**/**  //   *******  /**          │        │               **********/** **/**/**////////**             │
                                                 │          /**      /**   /**/**   ** **////**  /**          │        │              /**//////**/**** //****       /**             │
                                                 │          /********//****** //***** //******** ***          │        │              /**     /**/**/   ///** ********              │
                                                 │          ////////  //////   /////   //////// ///           │        │              //      // //       // ////////               │
                                                 └────────────────────────────────────────────────────────────┘        └────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────┐           ┌────────────────────────┐        ┌────────────────────────┐                                         ┌─ ── ── ── ── ── ── ── ── ──
│app1/docker-compose.yml              │           │         Proxy          │        │          web           │                                         │          Instance          │
│                                     │           │                        │        │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │         ┌────────────────────────┐      │ ┌────────────────────────┐ │
│web:                                 │         ┌─┴──┐                   ┌─┴─┐    ┌─┴─┐   ┌──┐               │         │          ELB           │        │          web           │
│  ports:                             │         │9000├───────────────────▶rnd├────▶rnd├───▶80│   nginx     │ │┌────────┴─────────────────┐      │      │ │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │ │
│    - 9000:80                        │         └─┬──┘                   └─┬─┘    └─┬─┘   └──┘               ││https://app1.elb.com:9000 │    ┌─┴─┐    ├─┴─┐   ┌──┐               │ │
│  labels:                            │           │                        │        │      └ ─ ─ ─ ─ ─ ─ ─ ┘ ││       (port req'd)       │────▶rnd├────▶rnd│───▶80│   nginx     │ │
│    - convox.port.9000.protocol=https│           │                        │        │                        │└────────┬─────────────────┘    └─┬─┘    ├─┬─┘   └──┘               │ │
└─────────────────────────────────────┘           └────────────────────────┘        └────────────────────────┘         │                        │      │ │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │ │
                                                                                                                       │                        │        │                        │
                                                                                                                       └────────────────────────┘      │ └────────────────────────┘ │
                                                                                                                                                       └ ── ── ── ── ── ── ── ── ── ┘

                                                                                                                                                       ┌─ ── ── ── ── ── ── ── ── ──
                                                                                                                                                       │          Instance          │
┌─────────────────────────────────────┐           ┌────────────────────────┐        ┌────────────────────────┐         ┌────────────────────────┐      │ ┌────────────────────────┐ │
│app2/docker-compose.yml              │           │         Proxy          │        │          web           │         │          ELB           │        │          web           │
│                                     │           │                        │        │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │┌────────┴─────────────────┐      │      │ │      ┌ ─ ─ ─ ─ ─ ─ ─ ┐ │ │
│web:                                 │         ┌─┴──┐                   ┌─┴─┐    ┌─┴─┐   ┌──┐               ││https://app2.elb.com:9000 │    ┌─┴─┐    ├─┴─┐   ┌──┐               │ │
│  ports:                             │         │9001├───────────────────▶rnd├────▶rnd├───▶80│   nginx     │ ││       (port req'd)       ├────▶rnd├────▶rnd│───▶80│   nginx     │ │
│    - 9001:80                        │         └─┬──┘                   └─┬─┘    └─┬─┘   └──┘               │└────────┬─────────────────┘    └─┬─┘    ├─┬─┘   └──┘               │ │
│  labels:                            │           │                        │        │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │         │                        │      │ │      └ ─ ─ ─ ─ ─ ─ ─ ┘ │ │
│    - convox.port.9001.protocol=https│           │                        │        │                        │         │                        │        │                        │
└─────────────────────────────────────┘           └────────────────────────┘        └────────────────────────┘         └────────────────────────┘      │ └────────────────────────┘ │
                                                                                                                                                       └ ── ── ── ── ── ── ── ── ── ┘


                                                            ┌─────────────────────────────────────┐                                ┌─────────────────────────────────────┐
                                                            │                                     │                                │                                     │
                                                            │               #     #               │                                │               #     #               │
                                                            │                #   #                │                                │               ##    #               │
                                                            │                 # #                 │                                │               # #   #               │
                                                            │                  #                  │                                │               #  #  #               │
                                                            │                  #                  │                                │               #   # #               │
                                                            │                  #                  │                                │               #    ##               │
                                                            │                  #                  │                                │               #     #               │
                                                            │                                     │                                │                                     │
                                                            └─────────────────────────────────────┘                                └─────────────────────────────────────┘
```

Not sure if that helps clarify our use case - I'll do some more thinking on possible solutions and follow up tomorrow.

**Comment by ddollar** [comment]
Proposed solution: #661

**Comment by ddollar** [comment]
Closed by #662
